### DISCUSSION TITLE
Extract the component ID of a vector

### DISCUSSION THREAD

**Discussion by QuangDom** [idaholab/moose#19201] [Q&A General] [answered]
Hi, 
I am looking for a method to extract the component id of a vector. For example, during the loop of _i in the method VectorIntegratedBC::computeResidual(). How can I know _i is belong to which component of the vector?

```
VectorIntegratedBC::computeResidual()
{
  prepareVectorTag(_assembly, _var.number());
  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
    for (_i = 0; _i < _test.size(); _i++)
    {
      Real residual = _JxW[_qp] * _coord[_qp] * computeQpResidual();
      _local_re(_i) += residual;
    }
  accumulateTaggedLocalResidual();
}
```

**Comment by GiudGiud** [comment, marked as answer]
_i should be the component index here. 
You can look at the computeQpResidual routine to convince yourself

Guillaume

**Reply by doquang** [reply]
Hi @GiudGiud,

In my case _i loops from 0..8. Where, the vector is a 2D-vector of QUAD4 element. Therefore, I would like to know which _i is x-component, and which _i is y-component.

Thanks in advance,

**Reply by GiudGiud** [reply]
oh true this is not as I thought. 
Ok so this BC is built so that the residual is (usually) the result vector product, multiplied by the value of a vector. So all the components of the variable are involved in computing the residual.
See https://github.com/idaholab/moose/blob/next/framework/src/bcs/VectorPenaltyDirichletBC.C for an example in framework/
and https://github.com/idaholab/moose/blob/next/modules/navier_stokes/src/bcs/INSADMomentumNoBCBC.C for an example in the context of fluid flow. 

You should not have to modify computeResidual btw. Only the computeQpResidual routines should need to be provided.
